### PR TITLE
Introduce a way to add protoc options to protobuf definitions

### PR DESCRIPTION
**Pull request by daddykotex**
The MetadataProcessor takes a model and extract a per namespace map of options. Before rendering a namespace we retrieve the options for a namespace and add them to the CompilationUnit.

Encoded as follows:

1. we use smithy metadata proto_options
2. we use an array of objects so that multiple proto_options can be concatenated
3. the options are defined per namespace
4. the options are String -> String

Regarding 4), I chose this encoding over String -> Any(bool, string, float, integer, identifier) because of the identifier.

You can do `option thing = "that";` but you can also do: `option thing = THAT;` where THAT would be a constant that protoc knows about.

To workaround the issue, the value is always a string, so if you want the former, you do `"thing": "\"that\""`, and if you want the latter you do: `"thing": "THAT"`. As such, if you want a `bool` like `true`, you do `"thing": "true"`.

I'm open to revisit this if you have another proposition.

### CODE REVIEW

**Comment by lewisjkl** [comment]
> To workaround the issue, the value is always a string, so if you want the former, you do "thing": "\"that\"", and if you want the latter you do: "thing": "THAT". As such, if you want a bool like true, you do "thing": "true".
>
> I'm open to revisit this if you have another proposition.

I think that is a great way to go at it. Simple and does the job 👍

**Review by lewisjkl — COMMENTED** [review]
Looks awesome! Will you add some docs to the readme about this? At that point I think we are good to 🚀

**Comment by daddykotex** [comment]
> Looks awesome! Will you add some docs to the readme about this? At that point I think we are good to 🚀

yes, I forgot

File: README.md [hunk 1]
```diff
@@ -1229,3 +1232,56 @@ message TestInput {
...
+
+Individual protobuf definitions file (`.proto`) can contain _options_. We support this feature using Smithy's metadata attribute. 
+
+There are a few importing things to notice
+
+1. All options are defined under the metadata key `proto_options`
+2. The value is an array. This is because Smithy will concatenate the arrays if the model contains multiple entries
+3. Each entry of the array is an object where the keys are the namespace and the values are objects that represent the options
+4. Entries for other namespaces are ignored (for example, `demo` in the example below)
+5. The object that represent an option can only use `String` as value (see the example below). More detail below.
```
Comment (lewisjkl) [inline README.md:1246]:
```suggestion
5. The object that represents an option can only use `String` as value (see the example below). More detail below.
```

File: modules/transitive/src/closure/TransitiveModel.scala [hunk 2]
```diff
@@ -31,7 +31,8 @@ object TransitiveModel {
       model: Model,
       entryPoints: List[ShapeId],
       captureTraits: Boolean,
-      validateModel: Boolean = true
+      captureMetadata: Boolean,
```
Comment (lewisjkl) [inline modules/transitive/src/closure/TransitiveModel.scala:34]: Nice find!

**Review by lewisjkl — APPROVED** [review]
Great work 💯

**Comment by lewisjkl** [comment]
@daddykotex is there anything holding up the merge of this PR?

**Comment by daddykotex** [comment]
nope it's good to go

**Pull request merged**